src/cargo/ops/cargo_config.rs

```diff
@@ -0,0 +1,314 @@
+            ),
+        }
+    }
+    Ok(())
+}
+
+/// Checks for environment variables that might be used.
+fn maybe_env<'config>(
+    config: &'config Config,
+    key: &ConfigKey,
+    cv: &CV,
+) -> Option<Vec<(&'config String, &'config String)>> {
+    // Only fetching a table is unable to load env values. Leaf entries should
+    // work properly.
+    match cv {
+        CV::Table(_map, _def) => {}
+        _ => return None,
+    }
+    let mut env: Vec<_> = config
+        .env()
+        .iter()
+        .filter(|(env_key, _val)| env_key.starts_with(&format!("{}_", key.as_env_key())))
+        .collect();
+    env.sort_by_key(|x| x.0);
+    if env.is_empty() {
+        None
+    } else {
+        Some(env)
+    }
+}
+
+fn print_toml(config: &Config, opts: &GetOptions<'_>, key: &ConfigKey, cv: &CV) {
+    let origin = |def: &Definition| -> String {
+        if !opts.show_origin {
+            return "".to_string();
+        }
+        format!(" # {}", def)
+    };
+    match cv {
+        CV::Boolean(val, def) => drop_println!(config, "{} = {}{}", key, val, origin(&def)),
+        CV::Integer(val, def) => drop_println!(config, "{} = {}{}", key, val, origin(&def)),
+        CV::String(val, def) => drop_println!(
+            config,
+            "{} = {}{}",
+            key,
+            toml::to_string(&val).unwrap(),
+            origin(&def)
+        ),
+        CV::List(vals, _def) => {
+            if opts.show_origin {
+                drop_println!(config, "{} = [", key);
+                for (val, def) in vals {
+                    drop_println!(config, "    {}, # {}", toml::to_string(&val).unwrap(), def);
+                }
+                drop_println!(config, "]");
+            } else {
+                let vals: Vec<&String> = vals.iter().map(|x| &x.0).collect();
+                drop_println!(config, "{} = {}", key, toml::to_string(&vals).unwrap());
+            }
+        }
+        CV::Table(table, _def) => {
+            let mut key_vals: Vec<_> = table.into_iter().collect();
+            key_vals.sort_by(|a, b| a.0.cmp(&b.0));
+            for (table_key, val) in key_vals {
+                let mut subkey = key.clone();
+                // push or push_sensitive shouldn't matter here, since this is
+                // not dealing with environment variables.
+                subkey.push(&table_key);
+                print_toml(config, opts, &subkey, val);
+            }
+        }
+    }
+}
+
+fn print_toml_env(config: &Config, env: &[(&String, &String)]) {
+    drop_println!(
+        config,
+        "# The following environment variables may affect the loaded values."
+    );
+    for (env_key, env_value) in env {
+        let val = shell_escape::escape(Cow::Borrowed(env_value));
+        drop_println!(config, "# {}={}", env_key, val);
+    }
+}
+
+fn print_json(config: &Config, key: &ConfigKey, cv: &CV, include_key: bool) {
+    let json_value = if key.is_root() || !include_key {
+        match cv {
```

Comment:
Could this be deduplicated with `json_add` below where there's a function from `CV` to a json value?